packages/frontend/amp/components/elements/find-adslots.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * Find ad slots
+ *
+ * Returns a list of indexes *after* which ads should be directly
+ * placed.
+ *
+ * Ads are placed:
+ *
+ * * sufficiently far from other ads (MIN_CHAR_BUFFER characters away)
+ * * sufficiently far from non-text (p) elements (IMG_BUFFER_[FWD|BWD])
+ * * non adjacent to small (SMALL_PARA_CHARS) paragraphs
+ *
+ * These tests apply forwards and backwards, though in the case of
+ * non-text buffers, the values differ.
+ *
+ * Where the above tests are met, we say there is adequate 'buffer'
+ * around the ad and it can be placed.
+ */
+
+export const AD_LIMIT = 8;
+export const SMALL_PARA_CHARS = 50;
+export const MIN_CHAR_BUFFER = 700;
+const IMG_BUFFER_FWD = 300; // really any non-p element type
+const IMG_BUFFER_BWD = 200;
+
+const isTextElement = (e: CAPIElement): boolean => {
+    return e._type === 'model.dotcomrendering.pageElements.TextBlockElement';
+};
+
+const getElementLength = (element: CAPIElement): number => {
+    switch (element._type) {
+        case 'model.dotcomrendering.pageElements.TextBlockElement':
+            return element.html.length;
```

Comment:
Does it calculate the length each time and how expensive is that? In the Scala, the HTML wrapping library calculates the length once I think which makes this okay, but should we memoise this or something ourselves here?

Reply:
I think this deserves eventual profiling to see if it's a problem.